src/format/strftime.rs

```diff
@@ -169,18 +174,93 @@ pub struct StrftimeItems<'a> {
     /// If the current specifier is composed of multiple formatting items (e.g. `%+`),
     /// parser refers to the statically reconstructed slice of them.
     /// If `recons` is not empty they have to be returned earlier than the `remainder`.
+    #[cfg(feature = "locales")]
+    recons: Vec<Item<'a>>,
+    #[cfg(not(feature = "locales"))]
     recons: &'static [Item<'static>],
```

Comment:
why does this need to be dynamically sized now? 

Ideally that shouldn't happen, it will cause a new vec to be allocated every iteration of the parse loop. If the vecs are definitely required then we need to find a way to re-use them, instead of creating new ones. It would be better to do `self.recons.clear(); self.recons.extend(...);` in all the places below where new ones are created. 

But I would like to be sure that that's actually necessary, it doesn't seem like it should be but I could definitely be missing something!

Reply:
haha :grin: I don't spontaneously think from that perspective :thinking: but I guess you're right, there are a lot of allocations in there.

It's for the date, time and date&time formats. They are dependent on the locale. Example [here](https://docs.rs/pure-rust-locales/0.5.2/pure_rust_locales/en_GB/LC_TIME/constant.D_FMT.html) in British UK where the day is first, then the month, then the year). And yes it is absolutely necessary.

These strings needs to be parsed first, which I do here: https://github.com/chronotope/chrono/blob/ea3d20e11e5c3e0420a04d9b88578aa6944bf08a/src/format/strftime.rs#L249-L251

Though these are constants so maybe there is a way to load them statically? Like with lazy_static? Also the length of the parsed iterator may vary per locale. I'm not sure this is a great idea though, it's a lot of memory used, there are a lot of locales.

Reply:
I think I got an idea of an alternative solution after looking at other crates.

Maybe we can split chrono into 2 crates: 1 for internals, 1 for chrono itself (or make a fake internals crate like serde does). Then I can probably build new `const` that use the internal representation of chrono  (`Item` and the other internal objects of chrono) using build.rs.

This would be like a conversion from the format of pure-rust-locales to chrono. Since they are const and they would be used only once, there would be no impact on memory and minimal impact on the binary size (correct me if I'm wrong).

Reply:
> Then I can probably build new const that use the internal representation of chrono (Item and the other internal objects of chrono) using build.rs.
>
> This would be like a conversion from the format of pure-rust-locales to chrono. Since they are const and they would be used only once, there would be no impact on memory and minimal impact on the binary size (correct me if I'm wrong).

Is it possible to do this in an internal private module in chrono? I agree that it seems like this should work.

That said, I'd be happy to consider that a future optimization, if you feel like it we can land the change to chrono without that and follow up later. If we go that route I'd definitely appreciate an issue explaining what you think the right way to do it is, so that we know where to go.

That said, there are some things that can be done without that constification to reduce both code complexity and allocations:

* For being able to read the code, create a single type alias that is the only place that
  we need to do the `cfg(locales)` (not sure about the exact syntax for the lifetime off
  the top of my head):

  ```rust
  #[cfg(feature = "locales")]
  type Fmt<'a> = Vec<Item<'a>>;
  #[cfg(not(feature = "locales"))]
  Type Fmt = &'static [Item<'static>]
  ```

  which allows the `StrftimeItems` to not have any cfg's inside of it.

* Maybe make a pair of cfg-alterate functions that generate the `Fmt` types, instead of
  needing to put the `cfg` in so many of the functions.
* That also makes it easier to centralize the logic of replacing the vec, so that it's
  either a strict overwrite or a clear + extended.

Reply:
> Is it possible to do this in an internal private module in chrono? I agree that it seems like this should work.

I believe it is **possible**. I gave a quick try to see how it goes and it seems it's going to be difficult: I didn't realize this but the format parser (parsing "%Y %d %m" to internal objects) is in the same module than parsing the date and time according to a format (`Utc.datetime_from_str("Fri Nov 28 12:00:09", "%a %b %e %T %Y"`). I'm not sure how to circumvent this issue without moving the modules and the components around... especially with the 1.13.

> I'd be happy to consider that a future optimization, if you feel like it we can land the change to chrono without that and follow up later.

If we go for the Vec for now we can release this as a new feature.

I really want to do the consts but I think the change is going to be intense (but it is definitely possible). If you think the implementation with clear&extend is good for now, then I will go for that for now so people can already enjoy the localization.

Reply:
done